parlai/agents/hugging_face/dict.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from parlai.core.dict import DictionaryAgent
+
+try:
+    from transformers import GPT2Tokenizer
+except ImportError as e:
+    raise e(
+        'Need to install Hugging Face transformers repository. '
+        'Try `pip install transformers`.'
+    )
+from abc import ABC, abstractmethod
+
+SPECIAL_TOKENS = {
+    "bos_token": "<bos>",
+    "eos_token": "<eos>",
+    "pad_token": "<pad>",
+}
+
+
+class HuggingFaceDictionaryAgent(DictionaryAgent, ABC):
+    """
+    Use Hugging Face tokenizers.
+    """
+
+    def __init__(self, opt):
+        super().__init__(opt)
+        # initialize from vocab path
+        self.tokenizer = self.get_tokenizer(opt)
+        self.override_special_tokens(opt)
+
+    @abstractmethod
+    def get_tokenizer(self, opt):
+        """
+        Instantiate the HuggingFace tokenizer for your model.
+        """
+        pass
+
+    @abstractmethod
+    def override_special_tokens(opt):
+        """
+        Override the special tokens for your tokenizer.
+        """
+        pass
+
+    def txt2vec(self, text, vec_type=list):
+        tokens = self.tokenizer.tokenize(text)
+        tokens_id = self.tokenizer.convert_tokens_to_ids(tokens)
+        return tokens_id
+
+    def vec2txt(self, vec):
+        return self.tokenizer.decode(vec, clean_up_tokenization_spaces=True)
+
+    def act(self):
+        """
+        Dummy override.
+        """
+        return {}
+
+
+class Gpt2DictionaryAgent(HuggingFaceDictionaryAgent):
+    def get_tokenizer(self, opt):
+        """
+        Instantiate tokenizer.
+        """
+        model_sz = opt['gpt2_size']
+        fle_key = 'gpt2' if model_sz == 'small' else f'gpt2-{model_sz}'
+        return GPT2Tokenizer.from_pretrained(fle_key)
+
+    def _define_special_tokens(self, opt):
+        if opt['add_special_tokens']:
```

Comment:
do we want a flag, or do we want to make an opinionated decision? We know it doesn't work well if you don't fine tune on dialogue and that's fine.

Reply:
No, I want a flag. For finetuning, we need the special tokens. But in order to evaluate teh model straight out of the box, we cannot add them (other wise PPL --> infinity)